micro-rdk-nmea/src/parse_helpers/parsers.rs

```diff
@@ -11,40 +11,42 @@ use super::{
 /// Cursor that consumes can consume bytes from a data vector by bit size.
 pub struct DataCursor {
     data: Vec<u8>,
+    // the amount of bits by which the previously read field overflowed into
+    // the first byte
+    bit_offset: usize,
 }
 
 impl DataCursor {
     pub fn new(data: Vec<u8>) -> Self {
-        DataCursor { data }
+        DataCursor {
+            data,
+            bit_offset: 0,
+        }
     }
 
     pub fn read(&mut self, bit_size: usize) -> Result<Vec<u8>, NmeaParseError> {
-        if bit_size / 8 > self.data.len() {
+        let bits_to_read = bit_size + self.bit_offset;
+        if bits_to_read / 8 > self.data.len() {
             return Err(NmeaParseError::NotEnoughData);
         }
         let mut res = Vec::new();
-        res.extend(self.data.drain(..(bit_size / 8)));
-
-        // If the next bit_size takes us into the middle of a byte, then
-        // we want the next byte of the result to be shifted so as to ignore the remaining bits.
-        // The first byte of the data should have the bits that were included in the result
-        // set to zero.
-        //
-        // This is due to the way number fields in NMEA messages appear to be formatted
-        // from observation of successfully parsed data, which is that overflowing bits
-        // are read in MsB order, but the resulting bytes are read in Little-Endian.
-        //
-        // For example, let's say the data is [179, 152], and we want to read a u16 from a bit size of 12
-        //
-        // [179, 152] is 39091 in u16, reading the first 12 bits and ignoring the last 4
-        // should yield [10110011 10011000] => [10110011 00001001] => 2483 (in Little-Endian)
-        if bit_size % 8 != 0 {
-            res.push(self.data[0] >> (8 - (bit_size % 8)));
-            let mask: u8 = 255 >> (bit_size % 8);
-            if let Some(first_byte) = self.data.get_mut(0) {
-                *first_byte |= mask;
+        res.extend(self.data.drain(..(bits_to_read / 8)));
```

Comment:
OK, just confirming, that here we are intentionally "under" reading. If `bits_to_read` is 15, we actually only want to read one byte permanently, then manually push a munged value we compute out of the byte we didn't permanently read, so that we can set the cursors `bit_offset` and leave it pointing to the partially read byte. Do I have that right?

Reply:
that is correct